bitcoin/src/blockdata/script.rs

```diff
@@ -1222,6 +1232,101 @@ mod test {
         script = script.push_opcode(OP_CHECKSIG); comp.push(0xACu8); assert_eq!(&script[..], &comp[..]);
     }
 
+    #[test]
+    fn test_p2pk_pubkey_bytes() {
+        // happy path
+        // A valid key and a valid script should return Some and equal the key passed
+        let key_str = "0411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3";
+        let key = PublicKey::from_str(key_str).unwrap();
+        let p2pk = Script::builder().push_key(&key).push_opcode(OP_CHECKSIG).into_script();
+        let actual = p2pk.p2pk_pubkey_bytes().unwrap();
+        assert_eq!(actual.to_vec(), key.to_bytes());
+
+        // invalid script checks
+        // A script without an OP_CHECKSIG should return None
+        let no_checksig = Script::builder().push_key(&key).into_script();
+        assert_eq!(no_checksig.p2pk_pubkey_bytes(), None);
+        // an empty script should return None
+        let empty_script = Script::builder().into_script();
+        assert!(empty_script.p2pk_pubkey_bytes().is_none());
+        // scripts with no key should return None
+        let no_push_bytes = Script::builder().push_opcode(OP_CHECKSIG).into_script();
+        assert!(no_push_bytes.p2pk_pubkey_bytes().is_none());
+        // scripts with a different op code other than OP_CHECKSIG should return None
+        let different_op_code = Script::builder().push_key(&key).push_opcode(OP_NOP).into_script();
+        assert!(different_op_code.p2pk_pubkey_bytes().is_none());
+
+        // invalid key check
+        // Should return None if the key is incorrect size (key is 63 bytes) 
+        let malformed_key = "21032e58afe51f9ed8ad3cc7897f634d881fdbe49816429ded8156bebd2ffd1";
+        let invalid_p2pk_script = Script::builder()
+            .push_slice(malformed_key.as_bytes())
+            .push_opcode(OP_CHECKSIG)
+            .into_script();
+        assert!(invalid_p2pk_script.p2pk_pubkey_bytes().is_none());
+        // invalid key should return Some, p2pk_pubkey_bytes doesn't care if the key is valid
+        let malformed_key = "21032e58afe51f9ed8ad3cc7897f634d881fdbe49816429ded8156bebd2ffd1ux";
+        let invalid_key_script = Script::builder()
+            .push_slice(malformed_key.as_bytes())
```

Comment:
This tests may be confusing since it actually uses ASCII values rather than hex decoding. Even though, it's technically correct.

Reply:
I was having a hard time figuring out how to make/find an invalid key. IIRC it's possible to create a key outside the curve?? But wasn't able to find anything online minus the finite field on wikipedia which didn't answer my question, so I opted for the technically correct solution.

Reply:
I'm not sure you understood what I mean. The *value* is technically correct. But you're not decoding the hex, you're using the ASCII values which is OK but someone could read it and think "WTF, this looks like a bug" or think that decoding the value will give him some interesting example of point not being on the curve.

This would be more readable:

```rust
let malformed_key = hex!("3231303332653538616665353166396564386164336363373839376636333464383831666462653439383136343239646564383135366265626432666664317578");
        let invalid_key_script = Script::builder()
            .push_slice(&malformed_key)
```

Note that I've just encoded the value you came up with as hex. :)